File: client/frontend/src/services/schedulerService.ts
import { SchedulerType, SchedulerConfig } from '../types/scheduler'
import { Task } from '../types/task'
import axios from 'axios'

const API_BASE_URL = '/api/scheduler'

export interface ScheduledTask {
  id: string
  name: string
  startTime: Date
  endTime: Date
  estimatedTime: number
  priority: number
}

export interface ScheduleEvaluation {
  totalTime: number
  averageWaitingTime: number
  averageTurnaroundTime: number
  efficiency: number
}

export interface ScheduleResult {
  scheduledTasks: ScheduledTask[]
  evaluation: ScheduleEvaluation
}

class FCFSScheduler {
    schedule(tasks: Task[]): ScheduleResult[] {
        const sortedTasks = [...tasks].sort((a, b) => 
            new Date(a.createTime).getTime() - new Date(b.createTime).getTime()
        )
        
        let currentTime = new Date()
        return sortedTasks.map(task => {
            const startTime = new Date(currentTime)
            const duration = task.estimatedTime || 30 // 默认30分钟
            currentTime = new Date(currentTime.getTime() + duration * 60000)
            
            return {
                taskId: task.id,
                startTime,
                endTime: new Date(currentTime),
                actualDuration: duration
            }
        })
    }

    evaluate(tasks: Task[]): ScheduleEvaluation {
        const scheduleResults = this.schedule(tasks)
        const totalTime = scheduleResults.reduce((sum, result) => 
            sum + result.actualDuration, 0
        )
        
        let totalWaitingTime = 0
        let totalTurnaroundTime = 0
        
        scheduleResults.forEach((result, index) => {
            const task = tasks.find(t => t.id === result.taskId)!
            const waitingTime = (new Date(result.startTime).getTime() - 
                new Date(task.createTime).getTime()) / 60000
            const turnaroundTime = waitingTime + result.actualDuration
            
            totalWaitingTime += waitingTime
            totalTurnaroundTime += turnaroundTime
        })
        
        return {
            totalTime,
            averageWaitingTime: totalWaitingTime / tasks.length,
            averageTurnaroundTime: totalTurnaroundTime / tasks.length,
            efficiency: 1 - (totalWaitingTime / (totalTime * tasks.length))
        }
    }
}

class SJFScheduler {
    schedule(tasks: Task[]): ScheduleResult[] {
        const sortedTasks = [...tasks].sort((a, b) => 
            (a.estimatedTime || 30) - (b.estimatedTime || 30)
        )
        
        let currentTime = new Date()
        return sortedTasks.map(task => {
            const startTime = new Date(currentTime)
            const duration = task.estimatedTime || 30
            currentTime = new Date(currentTime.getTime() + duration * 60000)
            
            return {
                taskId: task.id,
                startTime,
                endTime: new Date(currentTime),
                actualDuration: duration
            }
        })
    }

    evaluate(tasks: Task[]): ScheduleEvaluation {
        const scheduleResults = this.schedule(tasks)
        const totalTime = scheduleResults.reduce((sum, result) => 
            sum + result.actualDuration, 0
        )
        
        let totalWaitingTime = 0
        let totalTurnaroundTime = 0
        
        scheduleResults.forEach((result, index) => {
            const task = tasks.find(t => t.id === result.taskId)!
            const waitingTime = (new Date(result.startTime).getTime() - 
                new Date(task.createTime).getTime()) / 60000
            const turnaroundTime = waitingTime + result.actualDuration
            
            totalWaitingTime += waitingTime
            totalTurnaroundTime += turnaroundTime
        })
        
        return {
            totalTime,
            averageWaitingTime: totalWaitingTime / tasks.length,
            averageTurnaroundTime: totalTurnaroundTime / tasks.length,
            efficiency: 1 - (totalWaitingTime / (totalTime * tasks.length))
        }
    }
}

class RRScheduler {
    private timeSlice: number

    constructor(timeSlice: number = 15) {
        this.timeSlice = timeSlice
    }

    schedule(tasks: Task[]): ScheduleResult[] {
        const results: ScheduleResult[] = []
        const taskQueue = [...tasks]
        const remainingTime = new Map<string, number>(
            tasks.map(task => [task.id, task.estimatedTime || 30])
        )
        
        let currentTime = new Date()
        
        while (taskQueue.length > 0) {
            const task = taskQueue.shift()!
            const remaining = remainingTime.get(task.id)!
            
            if (remaining > 0) {
                const startTime = new Date(currentTime)
                const duration = Math.min(remaining, this.timeSlice)
                currentTime = new Date(currentTime.getTime() + duration * 60000)
                
                results.push({
                    taskId: task.id,
                    startTime,
                    endTime: new Date(currentTime),
                    actualDuration: duration
                })
                
                remainingTime.set(task.id, remaining - duration)
                
                if (remaining - duration > 0) {
                    taskQueue.push(task)
                }
            }
        }
        
        return results
    }

    evaluate(tasks: Task[]): ScheduleEvaluation {
        const scheduleResults = this.schedule(tasks)
        const totalTime = scheduleResults.reduce((sum, result) => 
            sum + result.actualDuration, 0
        )
        
        let totalWaitingTime = 0
        let totalTurnaroundTime = 0
        
        scheduleResults.forEach((result, index) => {
            const task = tasks.find(t => t.id === result.taskId)!
            const waitingTime = (new Date(result.startTime).getTime() - 
                new Date(task.createTime).getTime()) / 60000
            const turnaroundTime = waitingTime + result.actualDuration
            
            totalWaitingTime += waitingTime
            totalTurnaroundTime += turnaroundTime
        })
        
        return {
            totalTime,
            averageWaitingTime: totalWaitingTime / tasks.length,
            averageTurnaroundTime: totalTurnaroundTime / tasks.length,
            efficiency: 1 - (totalWaitingTime / (totalTime * tasks.length))
        }
    }
}

export const schedulerService = {
    createScheduler(config: SchedulerConfig) {
        switch (config.type) {
            case SchedulerType.FCFS:
                return new FCFSScheduler()
            case SchedulerType.SJF:
                return new SJFScheduler()
            case SchedulerType.RR:
                return new RRScheduler(config.timeSlice)
            default:
                throw new Error('不支持的调度算法类型')
        }
    },

    async scheduleTasks(algorithm: string, tasks: Task[]): Promise<ScheduleResult> {
        try {
            const response = await axios.post(`${API_BASE_URL}/schedule`, {
                algorithm,
                tasks: tasks.map(task => ({
                    id: task.id,
                    name: task.name,
                    priority: task.priority,
                    estimatedTime: task.estimatedTime,
                    startTime: task.startTime,
                    endTime: task.endTime
                }))
            })
            return response.data
        } catch (error) {
            console.error('调度任务失败:', error)
            throw error
        }
    },

    async getAvailableAlgorithms(): Promise<string[]> {
        try {
            const response = await axios.get(`${API_BASE_URL}/schedule/algorithms`)
            return response.data
        } catch (error) {
            console.error('获取可用算法失败:', error)
            throw error
        }
    },

    async evaluateSchedule(tasks: ScheduledTask[]): Promise<ScheduleEvaluation> {
        try {
            const response = await axios.post(`${API_BASE_URL}/schedule/evaluate`, {
                tasks
            })
            return response.data
        } catch (error) {
            console.error('评估调度失败:', error)
            throw error
        }
    },

    async setAlgorithm(algorithm: string): Promise<void> {
        await axios.post(`${API_BASE_URL}/algorithm`, { algorithm })
    },

    async generateSchedule(): Promise<any> {
        const response = await axios.post(`${API_BASE_URL}/generate`)
        return response.data
    }
} 

File: client/backend/src/services/scheduler_service.py
from typing import List
from models.task import Task as TaskModel
from schemas.scheduler import ScheduleRequest, ScheduleResult, ScheduleEvaluation
from sqlalchemy.orm import Session
from datetime import datetime, timedelta

class Scheduler:
    def __init__(self):
        pass

    def schedule(self, tasks: List[TaskModel]) -> List[ScheduleResult]:
        raise NotImplementedError
    
    def evaluate(self, tasks: List[TaskModel]) -> ScheduleEvaluation:
        raise NotImplementedError

class FCFSScheduler(Scheduler):
    def schedule(self, tasks: List[TaskModel]) -> List[ScheduleResult]:
        results = []
        current_time = 0
        for task in tasks:
            if task.status == 'PENDING':
                results.append(ScheduleResult(
                    task_id=task.id,
                    start_time=current_time,
                    end_time=current_time + (task.estimated_time or 60),
                    actual_duration=task.estimated_time or 60
                ))
                current_time += task.estimated_time or 60
        return results
    
    def evaluate(self, tasks: List[TaskModel]) -> ScheduleEvaluation:
        total_time = 0
        waiting_time = 0
        turnaround_time = 0
        current_time = 0
        
        for task in tasks:
            if task.estimated_time:
                waiting_time += current_time
                total_time += task.estimated_time
                turnaround_time += current_time + task.estimated_time
                current_time += task.estimated_time
                
        return ScheduleEvaluation(
            total_time=total_time,
            average_waiting_time=waiting_time / len(tasks) if tasks else 0,
            average_turnaround_time=turnaround_time / len(tasks) if tasks else 0,
            efficiency=total_time / (current_time if current_time > 0 else 1)
        )

class SJFScheduler(Scheduler):
    def schedule(self, tasks: List[TaskModel]) -> List[ScheduleResult]:
        results = []
        current_time = 0
        sorted_tasks = sorted([t for t in tasks if t.status == 'PENDING'], key=lambda x: x.estimated_time or float('inf'))
        for task in sorted_tasks:
            results.append(ScheduleResult(
                task_id=task.id,
                start_time=current_time,
                end_time=current_time + (task.estimated_time or 60),
                actual_duration=task.estimated_time or 60
            ))
            current_time += task.estimated_time or 60
        return results
    
    def evaluate(self, tasks: List[TaskModel]) -> ScheduleEvaluation:
        total_time = 0
        waiting_time = 0
        turnaround_time = 0
        current_time = 0
        
        sorted_tasks = sorted([t for t in tasks if t.estimated_time], key=lambda x: x.estimated_time)
        for task in sorted_tasks:
            waiting_time += current_time
            total_time += task.estimated_time
            turnaround_time += current_time + task.estimated_time
            current_time += task.estimated_time
                
        return ScheduleEvaluation(
            total_time=total_time,
            average_waiting_time=waiting_time / len(tasks) if tasks else 0,
            average_turnaround_time=turnaround_time / len(tasks) if tasks else 0,
            efficiency=total_time / (current_time if current_time > 0 else 1)
        )

class RRScheduler(Scheduler):
    def __init__(self, time_slice: int = 10):
        self.time_slice = time_slice

    def schedule(self, tasks: List[TaskModel]) -> List[ScheduleResult]:
        results = []
        current_time = 0
        remaining_tasks = [(task, task.estimated_time or 60) for task in tasks if task.status == 'PENDING']

        while remaining_tasks:
            for i, (task, remaining_time) in enumerate(remaining_tasks[:]):
                if remaining_time <= self.time_slice:
                    results.append(ScheduleResult(
                        task_id=task.id,
                        start_time=current_time,
                        end_time=current_time + remaining_time,
                        actual_duration=remaining_time
                    ))
                    current_time += remaining_time
                    remaining_tasks.pop(i)
                else:
                    results.append(ScheduleResult(
                        task_id=task.id,
                        start_time=current_time,
                        end_time=current_time + self.time_slice,
                        actual_duration=self.time_slice
                    ))
                    current_time += self.time_slice
                    remaining_tasks[i] = (task, remaining_time - self.time_slice)

        return results
    
    def evaluate(self, tasks: List[TaskModel]) -> ScheduleEvaluation:
        total_time = 0
        waiting_time = 0
        turnaround_time = 0
        current_time = 0
        
        remaining_tasks = [(task, task.estimated_time or 60) for task in tasks if task.estimated_time]
        while remaining_tasks:
            for i, (task, remaining_time) in enumerate(remaining_tasks[:]):
                if remaining_time <= self.time_slice:
                    waiting_time += current_time
                    total_time += remaining_time
                    turnaround_time += current_time + remaining_time
                    current_time += remaining_time
                    remaining_tasks.pop(i)
                else:
                    waiting_time += current_time
                    total_time += self.time_slice
                    turnaround_time += current_time + self.time_slice
                    current_time += self.time_slice
                    remaining_tasks[i] = (task, remaining_time - self.time_slice)
                
        return ScheduleEvaluation(
            total_time=total_time,
            average_waiting_time=waiting_time / len(tasks) if tasks else 0,
            average_turnaround_time=turnaround_time / len(tasks) if tasks else 0,
            efficiency=total_time / (current_time if current_time > 0 else 1)
        )

def create_scheduler(algorithm: str) -> Scheduler:
    if algorithm == 'first-come-first-serve':
        return FCFSScheduler()
    elif algorithm == 'shortest-job-first':
        return SJFScheduler()
    elif algorithm == 'round-robin':
        return RRScheduler()
    else:
        raise ValueError(f"不支持的调度算法: {algorithm}")

def get_available_algorithms() -> List[str]:
    return ['first-come-first-serve', 'shortest-job-first', 'round-robin']

def schedule_tasks(db: Session, algorithm: str, tasks: List[TaskModel]) -> ScheduleResult:
    scheduler = create_scheduler(algorithm)
    scheduled_tasks = scheduler.schedule(tasks)
    evaluation = scheduler.evaluate(tasks)
    return ScheduleResult(
        scheduled_tasks=scheduled_tasks,
        evaluation=evaluation
    )

def evaluate_schedule(db: Session, tasks: List[TaskModel]) -> ScheduleEvaluation:
    scheduler = FCFSScheduler()  # 使用FCFS作为默认评估算法
    return scheduler.evaluate(tasks) 